main/src/library/MutDisjointSet.flix

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright 2023 Jakob Schneider Villumsen
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+pub enum MutDisjointSet[t: Type, r: Eff]({
+    rc = Region[r],
+    nodes = Vector[(t, MutDisjointSet.Node[t, r])]
+})
+
+mod MutDisjointSet {
+
+    pub enum Node[t: Type, r: Eff]({
+        rc = Region[r],
+        value = t,
+        parent = Ref[Option[Node[t, r]], r],
+        rank = Ref[Int32, r]
+    })
+
+    instance Eq[Node[t, r]] with Eq[t] {
+        pub def eq(x: Node[t, r], y: Node[t, r]): Bool = {
+            let Node.Node({ value = xval | _ }) = x;
+            let Node.Node({ value = yval | _ }) = y;
+            Eq.eq(xval, yval)
+        }
+    }
+
+    instance Order[Node[t, r]] with Order[t] {
+        pub def compare(x: Node[t, r], y: Node[t, r]): Comparison = {
+            let Node.Node({ value = xval | _ }) = x;
+            let Node.Node({ value = yval | _ }) = y;
+            Order.compare(xval, yval)
+        }
+    }
+
+    ///
+    /// Returns the node in the set `s` that contains `x`.
+    /// Returns `None` if `x` is not in `s`.
+    ///
+    pub def node(x: t, s: MutDisjointSet[t, r]): Option[Node[t, r]] with Eq[t] = {
+        let MutDisjointSet({ nodes | _ }) = s;
+        Vector.find(fst >> Eq.eq(x), nodes)
+        |> Option.map(snd)
+    }
+
+    ///
+    /// Returns a new collection of disjoint singletong sets containing
+    /// the elements of the iterable `m` in the region `rc`.
+    ///
+    pub def makeSet(rc: Region[r], m: m[t]): MutDisjointSet[t, r] \ r with Iterable[m], Order[t] = {
```

Comment:
I understand the point here, but it requires us to know the entire set ahead of time.

I think it would be better to use a `MutMap` which can be updated and which can find elements in logarithmic time.

Reply:
Ah okay, it seems the literature doesn't handle this case though. I've only seen examples of sets that are given in advance and where you want to answer questions about the structure of that set.

Reply:
But it makes sense right?

Reply:
Yes

Reply:
Maybe I'm just confused, but isn't it hard to make it disjoint without having the set ahead of time? If I want to add `42` i'd have to do a linear search to see if its in already?

Reply:
> Maybe I'm just confused, but isn't it hard to make it disjoint without having the set ahead of time? If I want to add `42` i'd have to do a linear search to see if its in already?

Correct! However, most pseudocode examples catch this example by stating `if x is not already in the forest then ...`

Reply:
and then you'd have to have a binary search tree or something on the side to make that performant. seems weird

Reply:
Just do a lookup in a table. Cost O(1) or O(log n).